src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java

```diff
@@ -206,14 +209,11 @@ public static class Execution extends AbstractStepExecutionImpl {
                 getContext().onFailure(error);
             } else {
                 LOGGER.info(() -> "Blocking " + k);
-                String c = Jenkins.XSTREAM.toXML(getContext());
-                synchronized (s) {
-                    s.contexts.put(k, c);
-                }
                 sync = false;
             }
             synchronized (s) {
-                s.started.add(k);
+                // Even if we completed immediately, set the state to StartedState so waitForSuccess knows to stop waiting.
```

Comment:
Why? I would expect the wait to just wait for the key to be present, with any value.

Also did you mean
```suggestion
                // Even if we completed immediately, set the state to StartedState so waitForStart knows to stop waiting.
```
?

Reply:
Because otherwise if you call `SemaphoreStep.success` or `SemaphoreStep.failure` and then `SemaphoreStep.waitForStart` _without even running a `Pipeline` that uses the `semaphore` step_, `waitForStart` will complete immediately.

Reply:
Basically, to really get this right, there needs to be at least 5 states, and we need to add a few more transitions that currently do not exist:
* No entry in the map 
* ImmediateFailure (the step will fail, if/when it starts)
* ImmediateSuccess (the step will succeed, if/when it starts)
* Started/Waiting/Blocked (the step has started, and has not yet completed)
* Finished (the step has finished, with some result) (does not currently exist, we'd need to add various transitions to this state)

And I do not think that we can introduce the `Finished` state without also introducing various behavior changes in edge cases that might affect existing tests. We'd definitely need a PCT run.